standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/TxnHandler.java

```diff
@@ -864,6 +867,98 @@ public void abortTxns(AbortTxnsRequest rqst) throws NoSuchTxnException, MetaExce
+    try {
+      if (sqlGenerator.getDbProduct() == MYSQL) {
+        stmt.execute("SET @@session.sql_mode=ANSI_QUOTES");
+      }
+      if (table == null) {
+        String s = sqlGenerator.addForUpdateClause(
+                "select \"DB_ID\" from \"DBS\"  where \"DBS\".\"NAME\" = ? " +
+                        " and \"DBS\".\"CTLG_NAME\" = ?");
+        List<String> params = Arrays.asList(db, catalog);
+        pst = sqlGenerator.prepareStmtWithParameters(dbConn, s, params);
+        LOG.debug("Going to execute query <" + s.replaceAll("\\?", "{}") + ">",
+                quoteString(db), quoteString(catalog));
+        rs = pst.executeQuery();
+        if (rs == null || !rs.next()) {
+          throw new MetaException(" db with name " + db + " does not exist in catalog " + catalog);
+        }
+        long dbId = rs.getLong(1);
+        rs = stmt.executeQuery("select PARAM_VALUE from DATABASE_PARAMS where PARAM_KEY = " +
+                "'repl.last.id' and DB_ID = " + dbId);
+        if (rs == null || !rs.next()) {
+          stmt.executeUpdate("insert into DATABASE_PARAMS values ( " + dbId +
+                  " , 'repl.last.id' , " + lastReplId + ")");
+        } else {
+          stmt.executeUpdate("update DATABASE_PARAMS set PARAM_VALUE = " + lastReplId + " where DB_ID = " + dbId +
+                  " and PARAM_KEY = 'repl.last.id'");
+        }
+      } else if (partList == null || partList.isEmpty()) {
+        String s = sqlGenerator.addForUpdateClause(
+            "select \"TBL_ID\" from \"TBLS\", \"DBS\"  where \"DBS\".\"NAME\" = ? " +
+            " and \"DBS\".\"CTLG_NAME\" = ? and \"TBLS\".\"TBL_NAME\" = ? and \"DBS\".\"DB_ID\" = \"TBLS\".\"DB_ID\"");
+        List<String> params = Arrays.asList(db, catalog, table);
+        pst = sqlGenerator.prepareStmtWithParameters(dbConn, s, params);
+        LOG.debug("Going to execute query <" + s.replaceAll("\\?", "{}") + ">",
+                quoteString(db), quoteString(catalog), quoteString(table));
+        rs = pst.executeQuery();
+        if (rs == null || !rs.next()) {
+          throw new MetaException(" table with name " + table + " does not exist in db " + catalog + "." + db);
+        }
+        long tblId = rs.getLong(1);
+        rs = stmt.executeQuery("select PARAM_VALUE from TABLE_PARAMS where PARAM_KEY = " +
+                "'repl.last.id' and TBL_ID = " + tblId);
+        if (rs == null || !rs.next()) {
+          stmt.executeUpdate("insert into TABLE_PARAMS values ( " + tblId +
+                  " , 'repl.last.id' , " + lastReplId + ")");
+        } else {
+          stmt.executeUpdate("update TABLE_PARAMS set PARAM_VALUE = " + lastReplId + " where TBL_ID = " + tblId +
+                  " and PARAM_KEY = 'repl.last.id'");
+        }
+      } else {
+        for (String part : partList) {
+          String s = sqlGenerator.addForUpdateClause(
+            "select \"PART_ID\" from \"PARTITIONS\", \"DBS\", \"TBLS\" where \"DBS\".\"NAME\" = ? " +
```

Comment:
Can append IN clause and get bunch of part Ids instead of getting it one by one.

Reply:
in clause can cause sql injection issue

Reply:
Shall refer CompactionTxnHandler.markCleaned for usage of TxnUtils.buildQueryWithINClauseStrings to handle sql injection case. We append "?" based on number of values in IN clause and set it accordingly based on returned counts.